Tutorials/WinForms/Classic/bind-the-data-grid-to-large-data-source.md

```diff
@@ -0,0 +1,294 @@
+* Add the **XPBindingSource** component to the Form and rename it to **CustomersBindingSource**.
+* Rebuild the project and set properties:
+    * ObjectClassInfo=DxSample.DataAccess.Customer
+    * DisplayableProperties=Oid;ContactName
+* Select the **lookUpEdit1** editor placed near the **Customer** label and open its smart-tag at the top-right corner. Open the combo box near the **Data Source** field, find the **CustomersBindingSource** component under the **Other Data Sources > EditOrderForm List Instances** node, and select it. 
+* Select the **ContactName** and **Oid** properties as the **Display Member** and **Value Member** if they were not selected automatically.
+* Open the code editor and change the code as follows
+    <details>
+        <summary>Click to expand</summary>   
+
+    ```csharp
+    using DevExpress.Xpo;
+    using DevExpress.Xpo.DB.Exceptions;
+    using DevExpress.XtraEditors;
+    using DxSample.DataAccess;
+    using System;
+    using System.Windows.Forms;
+
+    namespace DxSample.Forms {
+        public partial class EditOrderForm : XtraForm {
+            public EditOrderForm() {
+                InitializeComponent();
+            }
+            public EditOrderForm(int? orderId) : this() {
+                OrderID = orderId;
+            }
+            public int? OrderID { get; private set; }
+            protected UnitOfWork UnitOfWork { get; private set; }
+            private void EditCustomerForm_Load(object sender, EventArgs e) {
+                Reload();
+            }
+
+            private void Reload() {
+                UnitOfWork = new UnitOfWork();
+                if (OrderID.HasValue)
+                    OrderBindingSource.DataSource = UnitOfWork.GetObjectByKey<Order>(OrderID.Value);
+                else OrderBindingSource.DataSource = new Order(UnitOfWork);
+                CustomersBindingSource.DataSource = new XPCollection<Customer>(UnitOfWork);
+            }
+
+            private void btnSave_Click(object sender, EventArgs e) {
+                try {
+                    UnitOfWork.CommitChanges();
+                    OrderID = ((Order)OrderBindingSource.DataSource).Oid;
+                    Close();
+                } catch (LockingException) {
+                    XtraMessageBox.Show(this, "The record was modified or deleted by another user. Please click the Reload button and try again.", "XPO Tutorial", MessageBoxButtons.OK, MessageBoxIcon.Stop);
+                }
+            }
+
+            private void btnReload_Click(object sender, EventArgs e) {
+                Reload();
+            }
+        }
+    }
+    ```
+    </details>
+## Create the orders list Form
+* Select the **CustomersListForm** item in the Solution Explorer window and press the **Ctrl+C** keyaborad shortcut to copy the Form
+* Press the **Ctrl+V** keyboard shortcut to create a CustomersListForm copy. Rename it to **OrdersListForm**
```

Comment:
and here

Reply:
502bbc1903824c6adf6045f4cd6d973ff23e763b